Fixes potion data conversion from 1.8 to 1.9

joserobjr opened the pull request:
As I said on https://github.com/MylesIsCool/ViaVersion/issues/420

I've made this convertor when spigot 1.9 was released. I adapted it for ViaVersion.

This change will still use the existing potion data registration when available, so it will not cause any impact where it's already working.

MylesIsCool commented:
This converts items from 1.8 id to 1.9 string, but does nothing to 1.9 string to 1.8 id. I'm unsure how the server works here because technically the item data id it sent to the client will not match the id that the client sends back (eg. moving an item).

Otherwise I feel the switch block needs to be formatted better with using spacing and new lines for the variables.

Also splash is commented out?

joserobjr commented:
> It does convert back, 1.9 uses simple string keys for potion IDs, a normal fire resistance potion is only possible with "fire_resistance" key on 1.9, it's not possible to detect which 1.8's data value created that 1.9 key because 1.9 ids and 1.8 data values are in an one to many relationship. But it makes no difference if the damage is 2 or 12323 because potions are not stackable and they produce the exact same effect. So it's fine to convert 1.9 to 1.8 using the same data value as creative tabs.

However I tested stack with many items and different damage values using Essentials plugin, it worked perfectly fine.

> Otherwise I feel the switch block needs to be formatted better with using spacing and new lines for the variables.
It's organized by one line per ID, that's why I skipped the case 7 and its line. It's very easy to read as it's now. Are you sure that you want to break readability with line breaks, I can do that in few seconds but will reduce the code maintainability...

Matsv commented:
I see you removed the line
```
 if (POTION_ID_TO_NAME.containsKey((int) item.getData())) {
```
Have you tested what will happen if you send it without any data?

joserobjr commented:
I removed it because potionNameFromDamage() will always resolve any data value to it's corresponding 1.9 id. No data means 1.8's `0` that means 1.9's `water`. 1.9 also removed the unused potions and replaced them by Uncraftable Potion. All that information is available on http://minecraft.gamepedia.com/Data_values#Potions , it's working fine on my server.

My tests was to:
1. Summon 3 stacks of stacked potions with different data values but with same effect
2. Attempt to stack them: It stacks for 100ms but it's immediately updated to the expected result, which is to replace the cursor with the stack from the clicked slot
3. Spawn potions from creative tabs: Works as expected
4. Spawn removed potions like Stinky Potion (name bits 63), Rank Potion (55), Sparkling Potion (47), Clear Potion (7), etc: They all resolves correctly to Uncraftable Potion (`empty`)
5. Create potions using Bukkit's Potion API (which is deprecated because of 1.9 features but works): Worked fine.

I did not find any issues yet, this change is being used all the time on my server, if you want to try the IP is sabotador.com, just let me know your nick and I set you on creative mode.

MylesIsCool commented on src/main/java/us/myles/ViaVersion/protocols/protocol1_9to1_8/ItemRewriter.java:
```diff
@@ -245,10 +243,72 @@ public static void toClient(Item item) {
+            if(canEnhance && enhanced) {
+                id = "strong_" + id;
+            }
+            else if(canExtend && extended) {
```
Formatting here, should be on same line as the if, otherwise makes it hard to read

MylesIsCool commented on src/main/java/us/myles/ViaVersion/protocols/protocol1_9to1_8/ItemRewriter.java:
```diff
@@ -245,10 +243,72 @@ public static void toClient(Item item) {
+        boolean canExtend = true;
+
+        String id;
+        switch (effect) {
```
I'd prefer this formatted to the java standards,

http://www.oracle.com/technetwork/java/codeconventions-150003.pdf

MylesIsCool commented on src/main/java/us/myles/ViaVersion/protocols/protocol1_9to1_8/ItemRewriter.java:
```diff
@@ -245,10 +243,72 @@ public static void toClient(Item item) {
+
+            default:
+                canEnhance = false; canExtend = false;
+                switch (name) {
```
Same formatting here

MylesIsCool commented:
I've left a few comments in relation to the formatting, otherwise it seems like a good patch :)

Matsv commented:
Thank you for your pull request, I manually merged it to change the formatting for you. :smile:
